### PR TITLE
Minify Refactor #4: JavaScript

### PR DESCRIPTION
**Pull request by Tabrisrp**
Same structure as the CSS part, but the combine class is more complicated as it's also combining inline and 3rd party scripts except the ones that are excluded

### CODE REVIEW

File: inc/classes/optimization/JS/class-combine.php [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+	 *
+	 * @since 3.1
+	 * @author Remy Perona
+	 *
+	 * @param string $minify_url URL to insert.
+	 * @return bool
+	 */
+	protected function inject_combined_url( $minify_url ) {
+		try {
+			$this->crawler->filter( 'body' )->append( '<script src="' . $minify_url . '" data-minify="1" />' );
```
Comment (Screenfeed) [inline inc/classes/optimization/JS/class-combine.php]: `esc_url()` maybe?

File: inc/classes/optimization/JS/class-combine.php [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+	/**
+	 * Patterns in content excluded from being combined
+	 *
+	 * @since 3.1
+	 * @author Remy Perona
+	 *
+	 * @return array
+	 */
+	protected function is_excluded_inline_content() {
```
Comment (Screenfeed) [inline inc/classes/optimization/JS/class-combine.php]: Should be named `get_excluded_inline_content()` imho.

File: inc/classes/optimization/JS/class-combine.php [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+	/**
+	 * Patterns in URL excluded from being combined
+	 *
+	 * @since 3.1
+	 * @author Remy Perona
+	 *
+	 * @return array
+	 */
+	protected function is_excluded_external_file() {
```
Comment (Screenfeed) [inline inc/classes/optimization/JS/class-combine.php]: Should be named `get_excluded_external_file_path()` imho.

File: inc/classes/optimization/JS/class-minify.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		if ( ! $nodes ) {
+			return $this->crawler->saveHTML();
+		}
+
+		$nodes->each( function( \Wa72\HtmlPageDom\HtmlPageCrawler $node, $i ) {
+			global $wp_scripts;
+
+			$src = $node->attr( 'src' );
+
+			if ( preg_match( '/(?:-|\.)min.js/iU', $src ) ) {
```
Comment (Screenfeed) [inline inc/classes/optimization/JS/class-minify.php]: Simpler: `/[-.]min\.js/iU`.

File: inc/classes/optimization/JS/class-minify.php [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+	 * @return string|bool The minify URL if successful, false otherwise
+	 */
+	protected function replace_url( $url ) {
+		if ( empty( $url ) ) {
+			return false;
+		}
+
+		$unique_id = md5( $url . $this->minify_key );
+		$filename  = preg_replace( '/\.(js)$/', '-' . $unique_id . '.js', ltrim( rocket_realpath( rocket_extract_url_component( $url, PHP_URL_PATH ) ), '/' ) );
```
Comment (Screenfeed) [inline inc/classes/optimization/JS/class-minify.php]: Parenthesis are unnecessary: `/\.js$/`.

File: inc/classes/optimization/JS/class-abstract-js-optimization.php [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	 *
+	 * @since 2.11
+	 * @author Remy Perona
+	 *
+	 * @param HtmlPageCrawler $node Node corresponding to a JS file.
+	 * @return bool True if it is a file excluded, false otherwise
+	 */
+	protected function is_minify_excluded_file( $node ) {
+		// File should not be minified.
+		if ( $node->attr( 'data-minify' ) || $node->attr( 'data-no-minify' ) ) {
```
[the remark this answers is not in the document]
Comment (Tabrisrp) [reply inc/classes/optimization/JS/class-abstract-js-optimization.php]: first one means the file is already minified, second means the file should not be

**Pull request merged**